File: lib/gurobi605/linux64/examples/python/sensitivity.py
#!/usr/bin/python

# Copyright 2014, Gurobi Optimization, Inc.

# A simple sensitivity analysis example which reads a MIP model
# from a file and solves it. Then each binary variable is set
# to 1-X, where X is its value in the optimal solution, and
# the impact on the objective function value is reported.

import sys
from gurobipy import *

if len(sys.argv) < 2:
    print('Usage: sensitivity.py filename')
    quit()

# Read and solve model

model = read(sys.argv[1])

if model.IsMIP == 0:
    print('Model is not a MIP')
    exit(0)

model.optimize()

if model.status != GRB.status.OPTIMAL:
    print('Optimization ended with status %d' % model.status)
    exit(0)

# Store the optimal solution

origObjVal = model.ObjVal
for v in model.getVars():
    v._origX = v.X

# Disable solver output for subsequent solves

model.params.OutputFlag = 0

# Iterate through unfixed, binary variables in model

for v in model.getVars():
    if (v.LB == 0 and v.UB == 1 \
        and (v.VType == GRB.BINARY or v.VType == GRB.INTEGER)):

        # Set variable to 1-X, where X is its value in optimal solution

        if v._origX < 0.5:
            v.LB = v.Start = 1
        else:
            v.UB = v.Start = 0

        # Update MIP start for the other variables

        for vv in model.getVars():
            if not vv.sameAs(v):
                vv.Start = vv._origX

        # Solve for new value and capture sensitivity information

        model.optimize()

        if model.status == GRB.OPTIMAL:
            print('Objective sensitivity for variable %s is %g' % \
                  (v.VarName, model.ObjVal - origObjVal))
        else:
            print('Objective sensitivity for variable %s is infinite' % \
                  v.VarName)

        # Restore the original variable bounds

        v.LB = 0
        v.UB = 1

File: lib/gurobi605/linux64/examples/java/Sensitivity.java
/* Copyright 2014, Gurobi Optimization, Inc. */

/* A simple sensitivity analysis example which reads a MIP model
   from a file and solves it. Then each binary variable is set
   to 1-X, where X is its value in the optimal solution, and
   the impact on the objective function value is reported.
*/

import gurobi.*;

public class Sensitivity {

  public static void main(String[] args) {

    if (args.length < 1) {
      System.out.println("Usage: java Sensitivity filename");
      System.exit(1);
    }

    try {

      // Create environment

      GRBEnv env = new GRBEnv();

      // Read and solve model

      GRBModel model = new GRBModel(env, args[0]);

      if (model.get(GRB.IntAttr.IsMIP) == 0) {
        System.out.println("Model is not a MIP");
        System.exit(1);
      }

      model.optimize();

      if (model.get(GRB.IntAttr.Status) != GRB.OPTIMAL) {
        System.out.println("Optimization ended with status "
            + model.get(GRB.IntAttr.Status));
        System.exit(1);
      }

      // Store the optimal solution

      double   origObjVal = model.get(GRB.DoubleAttr.ObjVal);
      GRBVar[] vars       = model.getVars();
      double[] origX      = model.get(GRB.DoubleAttr.X, vars);

      // Disable solver output for subsequent solves

      model.getEnv().set(GRB.IntParam.OutputFlag, 0);

      // Iterate through unfixed, binary variables in model

      for (int i = 0; i < vars.length; i++) {
        GRBVar v     = vars[i];
        char   vType = v.get(GRB.CharAttr.VType);

        if (v.get(GRB.DoubleAttr.LB) == 0 && v.get(GRB.DoubleAttr.UB) == 1
            && (vType == GRB.BINARY || vType == GRB.INTEGER)) {

          // Set variable to 1-X, where X is its value in optimal solution

          if (origX[i] < 0.5) {
            v.set(GRB.DoubleAttr.LB, 1.0);
            v.set(GRB.DoubleAttr.Start, 1.0);
          } else {
            v.set(GRB.DoubleAttr.UB, 0.0);
            v.set(GRB.DoubleAttr.Start, 0.0);
          }

          // Update MIP start for the other variables

          for (int j = 0; j < vars.length; j++) {
            if (j != i) {
              vars[j].set(GRB.DoubleAttr.Start, origX[j]);
            }
          }

          // Solve for new value and capture sensitivity information

          model.optimize();

          if (model.get(GRB.IntAttr.Status) == GRB.OPTIMAL) {
            System.out.println("Objective sensitivity for variable "
                + v.get(GRB.StringAttr.VarName) + " is "
                + (model.get(GRB.DoubleAttr.ObjVal) - origObjVal));
          } else {
            System.out.println("Objective sensitivity for variable "
                + v.get(GRB.StringAttr.VarName) + " is infinite");
          }

          // Restore the original variable bounds

          v.set(GRB.DoubleAttr.LB, 0.0);
          v.set(GRB.DoubleAttr.UB, 1.0);
        }
      }

      // Dispose of model and environment

      model.dispose();
      env.dispose();

    } catch (GRBException e) {
      System.out.println("Error code: " + e.getErrorCode());
      System.out.println(e.getMessage());
      e.printStackTrace();
    }
  }
}
